# bamUtil Module

## Logging:
Will be stored in: "log/bamUtil"

## Modules:
See invidual files for expanded explanations of their purpose and the manner in which they accomplish it.
* cleanBAM: Cleans the provided SAM/BAM, soft-clipping beyond-end-of-reference alignments and setting MAPQ to 0 for unmapped reads.
* filteredBAM: Filtering of a '.bam' file on read quality.
* fixmateBAM: Fill in mate coordinates, ISIZE, and mate related flags from a name-sorted alignment.
* indexBAM: Indexing of a '.bam' file.
* markdupBAM: Mark diplicates in a '.bam' file and produce metrics.
* mergeBAM: Merge a set of '.bam' files into a single '.bam' file.
* namesortBAM_biobambam: Produce a name sorted '.bam' file, via BIOBAMBAM.
* namesortBAM_samtools: Produce a name sorted '.bam' file, via SAMTOOLS. 
* rmdupBAM: Remove duplicates from a '.bam' file.
* sortBAM_biobambam: Sorting of a '.bam' file, via BIOBAMBAM.
* sortBAM_samtools: Sorting of a '.bam' file, via SAMTOOLS.
* **bamUtil_INCLUDE: Master file which includes the modules above.**

## Control Flags:
At times, there will need to exist submodules which only differ by the software they use and
the arguemnts that the respective software requires. To combined this into a single rule is possible
but, for the sake of simplicity and maximum cohesion, they have been seperated into different files.
As these two or more files will all produce the same output, Snakemake cannot otherwise make a choice
as to which of the options it must use. In this case, the following submodules exist with software 
variants. In order to override the default submodule software, users must set the following flags in 
the bamUtil section of the config.yaml file. If/elif/else conditionals were used, as such, the 
value flagged as "default" can be assumed to be contained within the else conditional.

YAML Flag | SubModules | Value required to set
:--------: | :--------: | :--------
SoftwareChoiceFlag_sortBAM | sortBAM_biobambam | biobambam
SoftwareChoiceFlag_sortBAM | sortBAM_samtools | samtools
SoftwareChoiceFlag_namesortBAM | namesortBAM_biobambam | biobambam
SoftwareChoiceFlag_namesortBAM | namesortBAM_samtools | samtools

## Global Directories:
Module | Argument | Default Value | Description
:--------: | :--------: | :--------: | :--------
All | bamDIR | bam | Directory to conatin the processed, symlinked, and name-normalized '.bam' and '.BAI' files.
alignBAM, bam2fastq | inputDIR | input | Directory to contain all files prior to starting pipeline (Reference and input files).
None | fastqDIR | fastq | Directory to contain the processed '.fastq' files.
markdup | metricsDIR | metrics | Directory to contain the processed '.metrics' files.
None | mpileupDIR | mpileup | Directory to contain the processed '.mpileup' files.
All | outputDIR | output | Directory to contain all files generated by the pipeline.

## Global Parameters:
Module | Argument | Default Value | Description
:--------: | :--------: | :--------: | :--------
All | shellCallFile | shellCalls.txt | File to which a copy of all shell calls is printed.
All | offCluster | False | Adds input redirection to end of shell calls as to capture STDERR when not on cluster.
None | fastqKEEP | False | Flag of 'True' or 'False' on retaining '.fastq' files after they are used in the pipeline.
None | inputPartList | inputPartList.txt | List containing the parts which comprise each sample.
None | refFILE | *Varies* | Absolute path to genomic reference file.
None | supportRefFILE | *Varies* | Absolute path to supporting genomic references files.
None | chrLIST | *Varies* | Full chromosomal list parse from one of the supporting genomic reference files.

## Module Specific Software:
Module | Argument | Default Value | Description
:--------: | :--------: | :--------: | :--------
alignBAM, indexBAM, filteredBAM, namesortBAM, rmdupBAM, sortBAM, bam2fastq | bamUtil_samtoolsProg | samtools | Path to Samtools executable.
cleanBAM, fixmateBAM, bam2fastq | bamUtil_picardProg | picard | Path to Picard executable.
alignBAM | bamUtil_bwaProg | bwa | Path to bwa executable.
fastqc | bamUtil_perlProg | perl | Path to perl executable.
namesortBAM, sortBAM | bamUtil_bamsortProg| bamsort | Path to bamsort executable.
markdupBAM |bamUtil_bammarkduplicates2Prog | bammarkduplicates2 | Path to bammarkduplicates2 executable.

## Module Specific Parameters:
Module | Argument | Default Value | Description
:--------: | :--------: | :--------: | :--------
All | bamDIR | bamUtil | Directory to contain the processed '.bam' files.
fastqc | fastqc | /genesis/clc/usr/fastqc-0.10.1/fastqc.pl | Perl Fastqc script to process a '.bam' file.
fastqc | nogroup | '' | Groups fastqc output based on certian inputs, alternatively set to: "--no-group"
cleanBAM, fixmateBAM | picardvalStringency | VALIDATION_STRINGENCY=LENIENT | Validation stringency for all SAM files read by this program.
cleanBAM, fixmateBAM | bamUtil_picardMaxRec | MAX_RECORDS_IN_RAM=5000000 | Specify the number of records stored in RAM before spilling to disk.
filteredBAM | bitFLAG | 512 | Must pass QC threshold.
indexBAM | fileTAG | _realigned_sorted_filtered_markdup | Tag facilitates bamUtil module processing of the '.bam' files in a specific order.
markdupBAM | compressLVL | level=-1 | Default @ -1 (zlib); Can go fast(1) or best(9) using gzip instead.
sortBAM | sortProgFLAG | biobambam | Flag to set whether Samtools(samtools) or Biobambam(biobambam) are used to sort.
