# vcfGenUtil_varScan Module

## Logging:
Will be stored in: "log/varScan"

## Modules:
See invidual files for expanded explanations of their purpose and the manner in which they accomplish it.
* mpileup2vcf_single: Generate single '.VCF' file from a single '.mpileup' file.
* mpileup2vcf_pair: Generate single chromosomal '.VCF' files from a pair of tumor-normal '.mpileup' files.
* copycall: Infer somatic copy number changes using data from matched tumor-normal pairs.
* copynum: Call variants and identify their somatic status
* conSeq: Generate concensus calls from a '.mpileup' file.
* **vcfGenUtil_varScan_INCLUDE: Master file which includes the submodules above.**

## Control Flags
None

## Global Directories:
Module | Argument | Default Value | Description
:--------: | :--------: | :--------: | :--------
None | bamDIR | bam | Directory to conatin the processed, symlinked, and name-normalized '.bam' and '.BAI' files.
None | inputDIR | input | Directory to contain all files prior to starting pipeline (Reference and input files).
None | fastqDIR | fastq | Directory to contain the processed '.fastq' files.
None | metricsDIR | metrics | Directory to contain the processed '.metrics' files.
mpileup2vcf* | mpileupDIR | mpileup | Directory to contain the processed '.mpileup' files.
All | outputDIR | output | Directory to contain all file generated by the pipeline.

## Global Parameters:
Module | Argument | Default Value | Description
:--------: | :--------: | :--------: | :--------
All | shellCallFile | shellCalls.txt | File to which a copy of all shell calls is printed.
All | offCluster | False | Adds input redirection to end of shell calls as to capture STDERR when not on cluster.
None | fastqKEEP | False | Flag of 'True' or 'False' on retaining '.fastq' files after they are used in the pipeline.
None | inputPartList | inputPartList.txt | List containing the parts which comprise each sample.
None | refFILE | *Varies* | Absolute path to genomic reference file.
None | supportRefFILE | *Varies* | Absolute path to supporting genomic references files.
None | chrLIST | *Varies* | Full chromosomal list parse from one of the supporting genomic reference files.

## Module Specific Software
Module | Argument | Default Value | Description
:--------: | :--------: | :--------: | :--------
mpileup2vcf* | vcfGenUtil_varScan_varScanProg | varScan | Program path.
mpileup2vcf* | vcfGenUtil_varScan_samtoolsProg | samtools | Program path.

## Module Specific Paramters:
Module |Argument | Default Value | Description
:--------: | :--------: | :--------: | :--------
All | vcfGenUtil_varScanDIRScanDIR | varScan | Log directory and core directory to store files.
mpileup2vcf* | minCOV | --min-coverage 20 | Minimum read depth at a position to make a call.
mpileup2vcf* | minREAD | --min-reads2 10 | Minimum supporting reads at a position to call variants.
mpileup2vcf* | minQUAL | --min-avg-qual 20 | Minimum base quality at a position to count a read.
mpileup2vcf* | minFREQ | --min-var-freq 0.01 | Minimum variant allele frequency threshold.
mpileup2vcf* | pVALUE | --p-value 0.05 | Default p-value threshold for calling variants.
mpileup2vcf* | strandFILT | --strand-filter 0 | Ignore variants with >90% support on one strand.
mpileup2vcf* | outVCF | --output-vcf 1 | If set to 1, outputs in VCF format.
mpileup2vcf_pair | minSTRAND | --min-strands2 0 | Minimum number of strands on which variant observed.
mpileup2vcf_pair | posVALID | --validation 0 | If set to 1, outputs all compared positions even if non-variant.
copynum | mapQSkip | -q 1 | Skip alignments with mapQ smaller than the INT.

## WARNINGS:
varScank.mk did not use the '-bh -F 512' flag when calculating indels for single samples, or when calcualting any tumor-normal pairs.
The snakemake equivalents all use the same code, so they all will use the '-bh -F 512' flags.
