# vcfAnnotate Module

## Logging:
Will be stored in: "log/vcfAnnotate"

## Modules:
See individual files for expanded explanations of their purpose and the manner in which they accomplish it.
* noncanonical: Canonical annotation of a '.VCF' file.
* canonical: Canonical annotation of a '.VCF' file.
* dbsnp: DbSnp annotation of a '.VCF' file.
* indel: INDEL annotation of a '.VCF' file.
* cosmic: Cosmic annotation of a '.VCF' file.
* **vcfAnnotate_INCLUDE: Annotation of '.VCF' files.***

## Control Flags:
None

## Global Directories:
Module | Argument | Default Value | Description
:--------: | :--------: | :--------: | :--------
None | bamDIR | bam | Directory to conatin the processed, symlinked, and name-normalized '.bam' and '.BAI' files.
None | inputDIR | input | Directory to contain all files prior to starting pipeline (Reference and input files).
None | fastqDIR | fastq | Directory to contain the processed '.fastq' files.
None | metricsDIR | metrics | Directory to contain the processed '.metrics' files.
None | mpileupDIR | mpileup | Directory to contain the processed '.mpileup' files.
All | outputDIR | output | Directory to contain all file generated by the pipeline.

## Global Parameters:
Module | Argument | Default Value | Description
:--------: | :--------: | :--------: | :--------
All | shellCallFile | shellCalls.txt | File to which a copy of all shell calls is printed.
All | offCluster | False | Adds input redirection to end of shell calls as to capture STDERR when not on cluster.
None | fastqKEEP | False | Flag of 'True' or 'False' on retaining '.fastq' files after they are used in the pipeline.
None | inputPartList | inputPartList.txt | List containing the parts which comprise each sample.
None | refFILE | *Varies* | Absolute path to genomic reference file.
None | supportRefFILE | *Varies* | Absolute path to supporting genomic references files.
None | chrLIST | *Varies* | Full chromosomal list parse from one of the supporting genomic reference files.

## Module Specific Software:
Module | Argument | Default Value | Description
:--------: | :--------: | :--------: | :--------
All | vcfAnnotate_javaProg | java -Xmx4G | Program path.
indel, cosmic, dbsnp | vcfAnnotate_snpSiftProg | -jar /genesis/extscratch/clc/usr/anaconda/4.3.0/envs/CentOS5-APR28-2/share/snpeff-4.1l-0/SnpSift.jar | Program path.
noncanonical, canonical | vcfAnnotate_snpEffProg | -jar /genesis/extscratch/clc/usr/anaconda/4.3.0/envs/CentOS5-APR28-2/share/snpeff-4.1l-0/snpEff.jar | Program path.
   

## Module Specific Paramters:
Module | Argument | Default Value | Description
:--------- | :--------: | :--------: | :--------
All | vcfAnnotate_javaProg | java -Xmx2G | Java version and the amount of memory it can use.
canonical | vcfAnnotate_snpEffProg | -jar ~/share/usr/anaconda/4.3.0/envs/CentOS5-Compatible/share/snpeff-4.1l-0/snpEff.jar | SnpEff program.
cosmic, indel, snp | vcfAnnotate_snpSiftProg | -jar ~/share/usr/anaconda/4.3.0/envs/CentOS5-Compatible/share/snpeff-4.1l-0/SnpSift.jar | SnpSift program.
All | vcfAnnotateDIR | vcfAnnotate | Directory to store annotated files.
All | annotateVerbose | -v | Turn on verbose output.
All | annotateID | -id | Turn on ID annotation.
All | annotateDbSnp | ~/share/references/dbsnp/dbsnp_137.hg19.vcf | Comparison database of SNPs.
canonical | snpEffGenomeVersion | GRCh37.72 | Version of reference genome to be used.
canonical | snpEffConfig | -c ~/share/usr/anaconda/4.3.0/envs/CentOS5-Compatible/share/snpeff-4.1l-0/snpEff.config | Config file to look up reference genome.
canonical | snpEffFormat | -i vcf | Output format.
canonical | snpEffHGVS | -hgvs | HGVS annotations for amino acid sub-field.
canonical | snpEffDownstream | -no-downstream | Do not show DOWNSTREAM changes.
canonical | snpEffIntergenic | -no-intergenic | Do not show INTERGENIC changes.
canonical | snpEffUpstream | -no-upstream | Do not show UPSTREAM changes.
cosmic | snpSiftCosmic | ~/share/references/cosmic/CosmicCodingMuts_v68.hg19.fixed.sort.vcf | Cosmic '.VCF' file database for comparison.
indel | knownIndel1000G | ~/share/references/1000g/1000G_phase1.indels.hg19.with_id.vcf | 1000G database file for comparison.
indel | knownIndelMills1000G | ~/share/references/1000g/Mills_and_1000G_gold_standard.indels.hg19.with_id.vcf | Mills_1000G database file for comparison.
